linkerd/proxy/transport/src/listen.rs

```
@@ -1,39 +1,111 @@
 use tokio::reactor;
+use tracing::trace;
 
-pub struct Listen {
-    local_addr: SocketAddr,
+/// A mockable source for address info, i.e., for tests.
+pub trait OrigDstAddr: Clone {
+    fn orig_dst_addr(&self, socket: &TcpStream) -> Option<SocketAddr>;
+}
+
+#[derive(Clone, Debug)]
+pub struct Bind<O: OrigDstAddr = NoOrigDstAddr> {
+    bind_addr: SocketAddr,
+    keepalive: Option<Duration>,
+    orig_dst_addr: O,
+}
+
+#[derive(Debug)]
+pub struct Listen<O: OrigDstAddr = NoOrigDstAddr> {
+    listen_addr: SocketAddr,
     keepalive: Option<Duration>,
+    orig_dst_addr: O,
     state: State,
 }
 
+pub type Connection = (Addrs, TcpStream);
+
+#[derive(Clone, Debug)]
+pub struct Addrs {
+    local: SocketAddr,
+    peer: SocketAddr,
+    orig_dst: Option<SocketAddr>,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct NoOrigDstAddr(());
+
+#[derive(Copy, Clone, Debug)]
+pub struct SysOrigDstAddr(());
+
+#[derive(Debug)]
 enum State {
     Init(Option<std::net::TcpListener>),
     Bound(tokio::net::TcpListener),
 }
 
-impl Listen {
-    pub fn bind(addr: SocketAddr, keepalive: Option<Duration>) -> std::io::Result<Self> {
-        let tcp = std::net::TcpListener::bind(addr)?;
-        let local_addr = tcp.local_addr()?;
-        Ok(Self {
-            local_addr,
+impl Bind {
+    pub fn new(bind_addr: SocketAddr, keepalive: Option<Duration>) -> Self {
+        Self {
+            bind_addr,
             keepalive,
-            state: State::Init(Some(tcp)),
-        })
+            orig_dst_addr: NoOrigDstAddr(()),
+        }
+    }
+}
+
+impl<O: OrigDstAddr> Bind<O> {
+    pub fn with_orig_dst_addrs_from<P: OrigDstAddr>(self, orig_dst_addr: P) -> Bind<P> {
+        Bind {
+            orig_dst_addr,
+            bind_addr: self.bind_addr,
+            keepalive: self.keepalive,
+        }
     }
 
-    pub fn local_addr(&self) -> SocketAddr {
-        self.local_addr
+    pub fn with_orig_dst_addrs_from_system(self) -> Bind<SysOrigDstAddr> {
```

nit: should this maybe fail fast (panic?) if not on Linux?